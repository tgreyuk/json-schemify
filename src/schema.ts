import * as fs from 'fs';
import * as path from 'path';
import { Json, Options } from 'types';
import { Schema } from './schema.interface';
import { parse } from './types/any';

/**
 * Returns a Json schema object
 * @param json A valid json structure
 * @param options Options to pass
 */
export function schemify(json: Json, options?: Options) {
  const schema: Schema = {
    ...(options && options.id && { $id: options.id }),
    $schema: 'http://json-schema.org/draft-07/schema#',
    ...(options && options.title && { title: options.title }),
    ...parse(json),
  };
  return schema;
}

/**
 * Write
 * @param json A valid json structure
 * @param filename The path of the output file to write to
 * @param options Options to pass
 */
export function writeSchema(json: Json, filepath: string, options?: Options) {
  const schema = schemify(json, options);
  if (!fs.existsSync(path.dirname(filepath))) {
    fs.mkdirSync(path.dirname(filepath));
  }
  if (options?.prettyPrint) {
    fs.writeFileSync(filepath, JSON.stringify(schema, null, 2));
  } else {
    fs.writeFileSync(filepath, JSON.stringify(schema));
  }
  if (!options?.silent) {
    console.log(`[json-schemify] Json schema written to ${filepath}`);
  }
}
