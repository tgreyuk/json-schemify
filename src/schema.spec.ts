import Ajv from 'ajv';
import addFormats from 'ajv-formats';
import * as fs from 'fs';
import * as tmp from 'tmp';
import { schemify, writeSchema } from './schema';

const mockJson = {
  title: 'Mock JSON',
  id: 100,
  date: '2013-10-21T13:28:06.419Z',
  fruits: ['apple', 'orange', 'pear'],
  mixed: ['stuff', null, 200],
  employers: [
    {
      name: 'John Doe',
      age: 21,
    },
    {
      name: 'John Doe',
      age: null,
    },
  ],
  person: {
    name: 'John Doe',
    age: 21,
  },
  coords: {
    latitude: 48.858093,
    longitude: 2.294694,
  },
};

beforeEach(() => {});

test('should declare JSON schema as draft-07', () => {
  const schema = schemify(mockJson);
  expect(schema.$schema).toEqual('http://json-schema.org/draft-07/schema#');
});

test('should declare a unique identifier if required', () => {
  const schema = schemify(mockJson, {
    id: 'http://shema/id.json',
  });
  expect(schema.$id).toEqual('http://shema/id.json');
});

test('should declare a title if required', () => {
  const schema = schemify(mockJson, { title: 'Mock Schema' });
  expect(schema.title).toEqual('Mock Schema');
});

test('should validate that the mock json validates against the generated schema', () => {
  const schema = schemify(mockJson, {
    id: 'http://shema/id.json',
    title: 'Mock Schema',
  });
  const ajv = new Ajv();
  addFormats(ajv);
  const valid = ajv.validate(schema, mockJson);
  if (ajv.errors) {
    console.error(ajv.errors);
  }
  expect(valid).toBeTruthy();
});

test('should write to file', () => {
  tmp.setGracefulCleanup();
  const tmpobj = tmp.dirSync();
  const jsonfile = tmpobj.name + '/schema.json';
  writeSchema(mockJson, jsonfile, {
    prettyPrint: true,
    silent: true,
  });
  expect(fs.readFileSync(jsonfile).toString()).toMatchSnapshot();
});
