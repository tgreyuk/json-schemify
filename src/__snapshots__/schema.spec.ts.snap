// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should write to file 1`] = `
"{
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"type\\": \\"object\\",
  \\"properties\\": {
    \\"title\\": {
      \\"type\\": \\"string\\"
    },
    \\"id\\": {
      \\"type\\": \\"integer\\"
    },
    \\"date\\": {
      \\"type\\": \\"string\\",
      \\"format\\": \\"date-time\\"
    },
    \\"fruits\\": {
      \\"type\\": \\"array\\",
      \\"items\\": {
        \\"type\\": \\"string\\"
      }
    },
    \\"mixed\\": {
      \\"type\\": \\"array\\",
      \\"items\\": {
        \\"oneOf\\": [
          {
            \\"type\\": \\"string\\"
          },
          {
            \\"type\\": \\"null\\"
          },
          {
            \\"type\\": \\"integer\\"
          }
        ]
      }
    },
    \\"employers\\": {
      \\"type\\": \\"array\\",
      \\"items\\": {
        \\"type\\": \\"object\\",
        \\"properties\\": {
          \\"name\\": {
            \\"type\\": \\"string\\"
          },
          \\"age\\": {
            \\"type\\": [
              \\"integer\\",
              \\"null\\"
            ]
          }
        }
      }
    },
    \\"person\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"name\\": {
          \\"type\\": \\"string\\"
        },
        \\"age\\": {
          \\"type\\": \\"integer\\"
        }
      }
    },
    \\"coords\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"latitude\\": {
          \\"type\\": \\"number\\"
        },
        \\"longitude\\": {
          \\"type\\": \\"number\\"
        }
      }
    }
  }
}"
`;
